File: frontend/src/components/benchmark/content/preferential_attachment/implementations/preferential_attachment.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class PreferentialAttachment
{
    private Dictionary<int, HashSet<int>> adjacency = new Dictionary<int, HashSet<int>>();
    private HashSet<int> nodes = new HashSet<int>();

    public void AddEdge(int u, int v)
    {
        if (!adjacency.ContainsKey(u))
            adjacency[u] = new HashSet<int>();
        if (!adjacency.ContainsKey(v))
            adjacency[v] = new HashSet<int>();

        adjacency[u].Add(v);
        adjacency[v].Add(u);
        nodes.Add(u);
        nodes.Add(v);
    }

    public void BuildFromEdges(List<(int, int)> edges)
    {
        foreach (var (u, v) in edges)
        {
            AddEdge(u, v);
        }
    }

    public double PreferentialAttachmentScore(int u, int v)
    {
        if (!adjacency.ContainsKey(u) || !adjacency.ContainsKey(v))
            return 0.0;

        return (double)adjacency[u].Count * adjacency[v].Count;
    }

    public List<((int, int), double)> ScoreAllPairs()
    {
        var scores = new List<((int, int), double)>();
        var nodeList = nodes.ToList();

        for (int i = 0; i < nodeList.Count; i++)
        {
            for (int j = i + 1; j < nodeList.Count; j++)
            {
                int u = nodeList[i];
                int v = nodeList[j];
                double score = PreferentialAttachmentScore(u, v);
                if (score > 0)
                {
                    scores.Add(((u, v), score));
                }
            }
        }

        return scores.OrderByDescending(x => x.Item2).ToList();
    }

    public List<((int, int), double)> TopKPredictions(int k = 10)
    {
        return ScoreAllPairs().Take(k).ToList();
    }
}

File: frontend/src/components/benchmark/content/preferential_attachment/implementations/preferential_attachment.rs
use std::collections::{HashMap, HashSet};

pub struct PreferentialAttachment {
    adjacency: HashMap<i32, HashSet<i32>>,
    nodes: HashSet<i32>,
}

impl PreferentialAttachment {
    pub fn new() -> Self {
        PreferentialAttachment {
            adjacency: HashMap::new(),
            nodes: HashSet::new(),
        }
    }

    pub fn add_edge(&mut self, u: i32, v: i32) {
        self.adjacency.entry(u).or_insert_with(HashSet::new).insert(v);
        self.adjacency.entry(v).or_insert_with(HashSet::new).insert(u);
        self.nodes.insert(u);
        self.nodes.insert(v);
    }

    pub fn build_from_edges(&mut self, edges: &[(i32, i32)]) {
        for &(u, v) in edges {
            self.add_edge(u, v);
        }
    }

    pub fn preferential_attachment_score(&self, u: i32, v: i32) -> f64 {
        if let (Some(neighbors_u), Some(neighbors_v)) = (self.adjacency.get(&u), self.adjacency.get(&v)) {
            return (neighbors_u.len() * neighbors_v.len()) as f64;
        }
        0.0
    }

    pub fn score_all_pairs(&self) -> Vec<((i32, i32), f64)> {
        let mut scores = Vec::new();
        let node_list: Vec<i32> = self.nodes.iter().cloned().collect();

        for i in 0..node_list.len() {
            for j in (i + 1)..node_list.len() {
                let u = node_list[i];
                let v = node_list[j];
                let score = self.preferential_attachment_score(u, v);
                if score > 0.0 {
                    scores.push(((u, v), score));
                }
            }
        }

        scores.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap());
        scores
    }

    pub fn top_k_predictions(&self, k: usize) -> Vec<((i32, i32), f64)> {
        let scores = self.score_all_pairs();
        scores.into_iter().take(k).collect()
    }
}